src/v/iceberg/values_json.cc

```diff
@@ -0,0 +1,699 @@
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/values_json.h"
+
+#include "iceberg/json_utils.h"
+
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_split.h>
+#include <boost/iostreams/device/array.hpp>
+#include <boost/iostreams/stream.hpp>
+#include <boost/range/irange.hpp>
+
+#include <charconv>
+#include <chrono>
+#include <iomanip>
+#include <ranges>
+#include <stdexcept>
+
+namespace {
+
+using namespace iceberg;
+
+std::string_view
+parse_string_view(const json::Value& data, std::string_view context = "") {
+    if (!data.IsString()) {
+        throw std::invalid_argument(fmt::format(
+          "Expected JSON string for {}, got {}", context, data.GetType()));
+    }
+    return std::string_view{data.GetString(), data.GetStringLength()};
+}
+
+iobuf hex_str_to_iobuf(std::string_view str) {
+    if (str.size() & 0x01ul) {
+        throw std::invalid_argument(
+          fmt::format("Expected even length hex string, got {}", str.size()));
+    }
+    bytes b;
+    b.reserve(str.size() / 2);
+    while (!str.empty()) {
+        auto sub = str.substr(0, 2);
+        uint8_t x{};
+        auto [ptr, ec] = std::from_chars(
+          sub.data(), sub.data() + sub.size(), x, 16);
+
+        if (ec != std::errc{}) {
+            throw std::invalid_argument(fmt::format(
+              "Failed to parse hex byte '{}' - ec: '{}'",
+              sub,
+              std::make_error_code(ec)));
+        }
+
+        b.push_back(static_cast<uint8_t>(x));
+        str = str.substr(2);
+    }
+    return bytes_to_iobuf(b);
+}
+
+// Convert a std::tm to a std::time_t (like mktime) but _without_ normalizing to
+// UTC. mktime assumes that the input tm is in your localtime, but we want to
+// treat these as already in UTC when parsing from JSON.
+// Code lifted more or less directly from a SO post from Howard Hinnant:
+// https://stackoverflow.com/a/56891776
+// The only change is to put all the arithmetic into longs to avoid
+// intermediate signed overflow.
+std::time_t to_time_t(const std::tm& tm) {
+    long y = tm.tm_year + 1900;
+    unsigned long m = tm.tm_mon + 1;
+    unsigned long d = tm.tm_mday;
+    y -= m <= 2;
+    const long era = (y >= 0 ? y : y - 399) / 400;
+    // vv [0, 399] vv
+    const auto yoe = static_cast<unsigned long>(y - era * 400);
+    // vv [0, 365] vv
+    const unsigned long doy = (153 * (m + (m > 2 ? -3 : 9)) + 2) / 5 + d - 1;
+    // vv [0, 146096] vv
+    const unsigned long doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
+    return (era * 146097 + static_cast<long>(doe) - 719468) * 86400
+           + static_cast<long>(tm.tm_hour * 3600)
+           + static_cast<long>(tm.tm_min * 60) + tm.tm_sec;
+}
+
+std::chrono::system_clock::time_point
+sv_to_time_point(std::string_view fmt, std::string_view str) {
```

Comment:
can absl::ParseTime help anything here?

I really find the stdlib stuff here really hard to work with, I think the absl stuff is generally cleaner and can convert to std::chrono

Reply:
not feeling compelled to learn a new parsing API, but absl::time does provide a sensible replacement for mktime, so I replaced the awful SO code w/ that.